### PR TITLE
polish template v-for error report

### LINKED ISSUE #164: Nested v-for breaks variable usage detection
* **ESLint Version:** 4.5.0
* **eslint-plugin-vue Version:** 3.12.0
* **Node Version:** 8.0.0
 
```js
module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:vue/recommended' // or 'plugin:vue/base'
  ],
}
```

Code:
```vue
<template>
  <div>
  <template v-for="item in dealItemList">
    <li v-for="snapItem in item.dealItemSnapList" :key="snapItem.id">
      123
    </li>
  </template>
  </div>
</template>
```

**What did you expect to happen?**

No error, since inner iterator `snapItem` depends on `item`, thus `snapItem.id` uses both inner and outer variables.

**What actually happened? Please include the actual, raw output from ESLint.**

Error.

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**
Fix #164, don't report error when nested v-for refers iterator.
Do report error when nested v-for doesn't refer iterator.

Supersede #167

### CODE REVIEW

File: lib/rules/valid-v-for.js [hunk 1]
```diff
@@ -19,34 +19,50 @@ const utils = require('../utils')
...
 
 /**
  * Check the given element about `v-bind:key` attributes.
  * @param {RuleContext} context The rule context to report.
  * @param {ASTNode} vFor The attribute node of `v-for` to check.
  * @param {ASTNode} element The element node to check.
+ * @param {boolean} isChild If element is a child of template[v-for].
  */
-function checkKey (context, vFor, element) {
+function checkKey (context, vFor, element, isChild) {
```
Comment (armano2) [inline lib/rules/valid-v-for.js]: checkKey is not always called with 4 parameters

Comment (HerringtonDarkholme) [reply lib/rules/valid-v-for.js]: labeled as optional

**Comment by mysticatea** [comment]
Thank you for contributing!

I think that this fix should be recursive. The `v-for` might be nested 3 times or more.
`vue-eslint-parser@2.0.0-beta.10` has the relationship between variables and references, I think we can use it. For example:

```js
/**
 * Find `VElement` node in the ancestor of the given node.
 * @param {Node} node The leaf node to find `VElement` node.
 * @returns {VElement|null} The found `VElement` node or null.
 */
function getElement (node) {
  while (node != null && node.type !== 'VElement') {
    node = node.parent
  }
  return node
}

/**
 * Check whether the given attribute is using the variables which are defined by `v-for` directives.
 * @param {VDirective} vFor The attribute node of `v-for` to check.
 * @param {VDirective} vBindKey The attribute node of `v-bind:key` to check. This can be a nested `v-for` directive.
 * @returns {boolean} `true` if the node is using the variables which are defined by `v-for` directives.
 */
function isUsingIterationVar (vFor, vBindKey) {
  if (vBindKey.value == null) {
    return false
  }
  const variables = vFor.parent.parent.variables

  return vBindKey.value.references.some(reference => {
    // Check only references which are resolved to `v-for` variables.
    if (reference.variable != null && reference.variable.kind === 'v-for') {
      // OK if the resolved variable is defined by the given `v-for` directive.
      if (variables.includes(reference.variable)) {
        return true
      }

      // Otherwise, the variable might be defined by a nested `v-for` directive.
      // Check whether the nested `v-for` directive is using the variable of the given `v-for` directive recursively.
      const element = getElement(reference.variable.id)
      const nestedVFor = element && utils.getDirective(element, 'for')
      return nestedVFor != null && isUsingIterationVar(vFor, nestedVFor)
    }
    return false
  })
}
```

**Comment by mysticatea** [comment]
Test cases:

valid

```
<template v-for="x in xs">
  <template v-for="y in x.ys">
    <li v-for="z in y.zs" :key="z.id">
      123
    </li>
  </template>
</template>
```

invalid

```
<template v-for="x in xs">
  <template v-for="y in a.ys">
    <li v-for="z in y.zs" :key="z.id">
      123
    </li>
  </template>
</template>
```

```
<template v-for="x in xs">
  <template v-for="y in x.ys">
    <li v-for="z in a.zs" :key="z.id">
      123
    </li>
  </template>
</template>
```

```
<template v-for="x in xs">
  <template v-for="y in x.ys">
    <li v-for="z in x.zs" :key="z.id">
      123
    </li>
  </template>
</template>
```

**Comment by HerringtonDarkholme** [comment]
Hmm, recursion might be too heavy. I will try some other solution.

**Comment by HerringtonDarkholme** [comment]
I have pushed a new checking method. The basic idea is if child has a v-for, only check if it has used parent iterator. If yes, we can delegate other tests to child v-for. If not, we need to apply old checking logic to descendant.

**Comment by HerringtonDarkholme** [comment]
Thanks!

**Pull request merged**